packages/engine/tests/experiment/mod.rs

```diff
@@ -41,16 +46,21 @@ fn load_manifest<P: AsRef<Path>>(project_path: P, language: Option<Language>) ->
 
     // We read the behaviors and datasets like loading a dependency
     let mut manifest = Manifest::from_dependency(project_path)
-        .wrap_err_lazy(|| format!("Could not load manifest from {project_path:?}"))?;
+        .wrap_err_lazy(|| format!("Could not load manifest from {project_path:?}"))
+        .provide_context(TestContext::ExperimentSetup)?;
```

Comment:
Actually, I think this is a good example. `from_dependency` gives us a `Report` with a proper error stack. We add a contextual message, that we couldn't load the manifest from the specified project path, and then we are changing the context to `ExperimentSetup`, which tells the caller, that the experiment setup is invalid. The outer function is about to load the manifest so IMO the context is pretty accurate. 

What message do you want to attach next to "Could not load..."? What additional information do you want to convey with a Provider, do you have something in mind?

Reply:
I don't disagree with anything you said there, I'm specifically asking about using an `enum` instead of a struct, that's all :) (And again I **don't dislike** the enum, it just feels different to how we'd talked about it)

In the comment above where I talk about function boundaries, I'm saying one might expect `Result<T, Report<ExperimentSetup>>` instead of `Result<T, Report<TestContext>>`? Does that kind of make sense?

Reply:
Yes, this is a good point. Especially when looking at popular crates (like `std`). Unless there is a good reason why multiple errors can happen (`io::Error`) it's always a struct. 

I gave it a try and also have an idea on how to solve this, but the main problem is, that all these errors currently happen in `run_test`. The nice solution would be to split `run_test` into `setup_experiment`, `run_experiment`, and `check_experiment_outputs`. However, we must not fail until we capture the results as we do all these cleanup logic and I currently don't want to rewrite the logic for running the tests. As the function is `async` it's even harder. I have put a TODO here but probably won't come back to this in the next days. adb44c8ebc6887b342107347ac6117ff3842bda1

Reply:
Cool, glad we're on the same page now. I think using an enum isn't a bad route here, and maybe we don't need to decide yet, I'm just wondering if the actual variations are things we want to capture as messages, spans, or something rather than changes to context 🤷 I don't know. We shouldn't view **error**s in other crates (like `std`) as being perfectly transferable here because `Context`s aren't necessarily errors :)

A TODO is good.